pkg/engine/executor/dataobjscan_predicate_test.go

```diff
@@ -649,3 +650,124 @@ func TestMapMetadataPredicate(t *testing.T) {
+		},
+		{
+			name: "not regex match filter",
+			expr: &physical.BinaryExpr{
+				Left:  &physical.ColumnExpr{Ref: types.ColumnRef{Column: "message", Type: types.ColumnTypeBuiltin}},
+				Right: physical.NewLiteral(`\d{4}-\d{2}-\d{2}`),
+				Op:    types.BinaryOpNotMatchRe,
+			},
+			expectedType: "logs.LogMessageFilterRowPredicate",
+		},
+		{
+			name: "pattern match filter",
+			expr: &physical.BinaryExpr{
+				Left:  &physical.ColumnExpr{Ref: types.ColumnRef{Column: "message", Type: types.ColumnTypeBuiltin}},
+				Right: physical.NewLiteral("<_> dataobj <_>"),
+				Op:    types.BinaryOpMatchPattern,
+			},
+			expectedErr: "unsupported binary operator (MATCH_PAT) for string predicate",
+		},
+		{
+			name: "not pattern match filter",
+			expr: &physical.BinaryExpr{
+				Left:  &physical.ColumnExpr{Ref: types.ColumnRef{Column: "message", Type: types.ColumnTypeBuiltin}},
+				Right: physical.NewLiteral("<_> dataobj <_>"),
+				Op:    types.BinaryOpNotMatchPattern,
+			},
+			expectedErr: "unsupported binary operator (NOT_MATCH_PAT) for string predicate",
+		},
+		{
+			name: "and filter",
+			expr: &physical.BinaryExpr{
+				Left: &physical.BinaryExpr{
+					Left:  &physical.ColumnExpr{Ref: types.ColumnRef{Column: "message", Type: types.ColumnTypeBuiltin}},
+					Right: physical.NewLiteral("foo"),
+					Op:    types.BinaryOpMatchSubstr,
+				},
+				Right: &physical.BinaryExpr{
+					Left:  &physical.ColumnExpr{Ref: types.ColumnRef{Column: "message", Type: types.ColumnTypeBuiltin}},
+					Right: physical.NewLiteral("bar"),
+					Op:    types.BinaryOpNotMatchSubstr,
+				},
+				Op: types.BinaryOpAnd,
+			},
+			expectedType: "logs.AndRowPredicate",
+		},
+		{
+			name: "or filter",
+			expr: &physical.BinaryExpr{
+				Left: &physical.BinaryExpr{
+					Left:  &physical.ColumnExpr{Ref: types.ColumnRef{Column: "message", Type: types.ColumnTypeBuiltin}},
+					Right: physical.NewLiteral("foo"),
+					Op:    types.BinaryOpMatchSubstr,
+				},
+				Right: &physical.BinaryExpr{
+					Left:  &physical.ColumnExpr{Ref: types.ColumnRef{Column: "message", Type: types.ColumnTypeBuiltin}},
+					Right: physical.NewLiteral("bar"),
+					Op:    types.BinaryOpNotMatchSubstr,
+				},
+				Op: types.BinaryOpOr,
+			},
+			expectedType: "logs.OrRowPredicate",
+		},
+		{
+			name: "not filter",
+			expr: &physical.UnaryExpr{
+				Left: &physical.BinaryExpr{
+					Left:  &physical.ColumnExpr{Ref: types.ColumnRef{Column: "message", Type: types.ColumnTypeBuiltin}},
+					Right: physical.NewLiteral("foo"),
+					Op:    types.BinaryOpMatchSubstr,
+				},
+				Op: types.UnaryOpNot,
+			},
+			expectedType: "logs.NotRowPredicate",
+		},
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			t.Logf("%s", tc.expr)
+			got, err := mapMessagePredicate(tc.expr)
+			if tc.expectedErr != "" {
+				require.ErrorContains(t, err, tc.expectedErr)
+			} else {
+				require.NoError(t, err)
+				require.Equal(t, tc.expectedType, fmt.Sprintf("%T", got))
```

Comment:
i guess we'll need a printer for predicates to write correctness tests. i had something for the inner predicates in a  [shelved PR](https://github.com/grafana/loki/pull/16846/), can revive as we invest more time on this